harness/determined/pytorch/README.md

```diff
@@ -0,0 +1,180 @@
+through.
+
+### The backward pass
+
+Once you have arrived at the loss you want to minimize, follow the the magic
+traceable autodiff path backward to calculate the gradient.  The gradients are
+left in-place on the tensors to which they apply.
+
+### `optimizer.step()`
+
+Gather up all of the gradents for the tensors this optimizer is responsible for
+tuning, and apply the `gradient * learning_rate` to each weight.  Now you have
+new weights, congratulations!
+
+
+## Horovod optimizer
+
+The horovod optimizer calculates gradient updates on each worker, then
+aggregates the updates across each worker before updating the weights in the
+model on each worker.
+
+### `loss.backward()`
+
+Horovod registers backward hooks on all of the `hvd_optimizer`'s parameter
+tensors; each backward pass starts an asynchronous allreduce of the gradient
+updates for parameters from that loss tensor.  The use of backward hooks makes
+this API-transparent to users.
+
+### `hvd_optimizer.synchronize()`
+
+Gather the results of the asynchronous allreduce of gradient updates between
+workers (don't apply any gradient updates).  This is useful if you need to
+process the gradients before applying them to the model weights, such as for
+gradient cliping or gradient scaling.
+
+### `hvd_optimizer.step()`
+
+First call .synchronize(), then apply the allreduced gradient updates by
+calling .step() on the wrapped optimizer.  This is the most vanilla way to use
+horovod.
+
+### `with hvd_optimizer.skip_synchronize():`
+
+Any call to `hvd_optimizer.step()` within this context manager will skip the
+.synchronize() step, which is useful if you already called
+`hvd_optimizer.synchronize()`.
+
+### `hvd_optimizer.backward_passes_per_step`
+
+Modifies the backward hooks to only start the asynchronous allreduce every
+`backward_passes_per_step` steps.  This value is useful for either gradient
+aggregation (mutliple batches per optimizer.step()) or if you plan on calling
+loss.backward() multiple times.
+
+Note that in Determined, the option named `backward_passes_per_step` is only
+for calling loss.backward() multiple times, and we multiply that value by the
+gradient aggregation frequency to get the `backward_passes_per_step` that we
+actually pass to Horovod.
+
+### optimizer.compression=hvd.Compression.fp16
+
+This is not AMP, and it does not behave like the GradScaler.  It just blindly
+casts gradient update tensors to fp16 for network communication and blindly
+casts them back to fp32 afterwards.
+
+
+## Gradient Aggregation:
+
+Gradient aggregation by a factor of N is just like mulitplying the batch size
+by N, except it saves on gpu memory because you process two batches and average
+their gradients before you apply the gradients.  This makes sense in
+distributed training because of the high communication cost of the allreduce.
+(There are some cases where it's not "just like multiplying batch size by N",
+such as with batch norm or if you are TODO:finish)
```

Comment:
I'm not sure what the remaining TODO is here.

Reply:
oh yeah, I think that turned out to be the detail about averaging the aggregated gradients.